library/alloc/src/alloc.rs

```diff
@@ -209,16 +209,14 @@ unsafe impl AllocRef for Global {
         );
 
         // SAFETY: `new_size` must be non-zero, which is checked in the match expression.
+        // If `new_size` is zero, then `old_size` has to be zero as well.
         // Other conditions must be upheld by the caller
```

Comment:
This is not documented though, is it? It least `grow` does not have a comment explaining the caller obligations.

Looks like `new_size >= old_size` would have to be among those.

Reply:
It's a safety constraint, that `new_size` must be greater than or equal to `old_size`. If `old_size` is non-zero (was checked just before), `new_size` can't be zero.

Reply:
I have a commit ready to be pushed, which cleans up the implementations and documentation of `AllocRef` a bit, so those things get more clear. The `unsafe` should be moved to line 217 instead. The match arm, where `layout.size() == 0` does not need `unsafe` at all.

Reply:
Right, my only comment is that this is an *undocumented* safety constraint -- at least when looking just at the grow function, and not at the entire file.

Reply:
However, I didn't want to change implementations much in this PR because of the perf-run.

Reply:
> But the usual comment explaining to the caller what they have to uphold is missing.

This is stated in the [safety-section of the trait](https://doc.rust-lang.org/nightly/std/alloc/trait.AllocRef.html#safety-2).

Reply:
Oh I see... in GH I couldn't see that this is a trait impl. Sorry.

Reply:
No Problem!
In any case #75657 should make this more readable if you only read the code. :)